File: DFS/135.Combination Sum/Solution.py
class Solution:
    """
    @param candidates: A list of integers
    @param target: An integer
    @return: A list of lists of integers
    """
    def combinationSum(self, candidates, target):
        # write your code here
        if not candidates or len(candidates) == 0:
            return [[]]

        candidates.sort()
        self.results = []
        self.target = target
        visited = [False for _ in range(len(candidates))]
        self._find_combination_sum(candidates, [], visited, 0)
        return self.results

    def _find_combination_sum(self, candidates, current_combination, visited, start):
        if sum(current_combination) == self.target:
            current_combination.sort()
            self.results.append(current_combination[:])
            return

        for i in range(start, len(candidates)):
            if sum(current_combination) + candidates[i] > self.target:
                break

            if i > 0 and candidates[i - 1] == candidates[i] and not visited[i - 1]:
                continue

            current_combination.append(candidates[i])
            visited[i] = True
            self._find_combination_sum(candidates, current_combination, visited, i)
            current_combination.pop()
            visited[i] = False

File: DFS/135.Combination Sum/Solution.java
public class Solution {
    /**
     * @param candidates: A list of integers
     * @param target: An integer
     * @return: A list of lists of integers
     */
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        // write your code here
        List<List<Integer>> results = new ArrayList<List<Integer>>();
        if (candidates == null || candidates.length == 0) {
            return results;
        }

        // Remove duplicates and sort the array
        int[] nums = removeDuplicateSort(candidates);

        helper(nums, 0, target, new ArrayList<Integer>(), results);
        return results;
    }

    private void helper(int[] nums,
                        int startIndex,
                        int target,
                        List<Integer> combination,
                        List<List<Integer>> results) {
        // Definition: recursively find correct combinations that sum up to target

        // Exit:
        if (target == 0) {
            results.add(new ArrayList<>(combination));
            return;
        }

        // Split:
        int length = nums.length;
        for (int i = startIndex; i < length; i++) {
            if (nums[i] <= target) {
                combination.add(nums[i]);
                helper(nums, i, target - nums[i], combination, results);
                combination.remove(combination.size() - 1);
            }
        }

    }

    // Remove Duplicate && Sorting
    private int[] removeDuplicateSort(int[] candidates) {
        Arrays.sort(candidates);
    
        int index = 0;
        int length = candidates.length;

        for (int i = 0; i < length; i++) {
            if (candidates[i] != candidates[index]) {
              candidates[++index] = candidates[i];
            }
        }

        int[] nums = new int[index + 1];

        for (int i = 0; i < index + 1; i++) {
            nums[i] = candidates[i];
        }

        return nums;
    }

}